File: 2024/dec04/C/aux04.c
#include "dec04.h"

bool check_horizontal(char list[][140], int row, int col) {
    if (col > COLUMNS-4) return false;

    char word[5]={0}; //it has to include '\0'
    for (int i=0; i<4; i++)
        word[i]=list[row][col+i];
    //printf("%s\n",word);
    return (!strcmp(word,"XMAS") || !strcmp(word,"SAMX"));
}

bool check_vertical(char list[][140], int row, int col) {
    if (row > ROWS-4) return false;
    
    char word[5]={0};
    for (int i=0; i<4; i++)
        word[i]=list[row+i][col];
    //printf("%s\n",word);
    return (!strcmp(word,"XMAS") || !strcmp(word,"SAMX"));
}

bool check_diagonal(char list[][140], int row, int col) {
    if (col > COLUMNS-4 || row > ROWS-4)
        return false;

    char word[5]={0};
    for (int i=0; i<4; i++)
        word[i]=list[row+i][col+i];
    //printf("%s\n",word);
    return (!strcmp(word,"XMAS") || !strcmp(word,"SAMX"));
}

bool check_anti_diagonal(char list[][140], int row, int col) {
    if (col < 3 || row > ROWS-4) return false;

    char word[5]={0};
    for (int i=0; i<4; i++)
        word[i]=list[row+i][col-i];
    //printf("%s\n",word);
    return (!strcmp(word,"XMAS") || !strcmp(word,"SAMX"));
}

bool x_mas(char list[][140], int row, int col) {
    // 1 . .    . . 3
    // . 2 .    . 2 .
    // . . 3    1 . .
    if (col > COLUMNS-3 || row > ROWS-3)
        return false;

    char word1[4]={0},word2[4]={0};
    for (int i=0; i<3; i++) {
        word1[i]=list[row+i][col+i];
        word2[i]=list[row+i][col+2-i];
    }
    bool w1=!(strcmp(word1,"MAS") && strcmp(word1,"SAM"));
    bool w2=!(strcmp(word2,"MAS") && strcmp(word2,"SAM"));

    return w1 && w2;
}

File: 2024/dec04/Python/aux04.py
inp = open('../input.txt', 'r')
lines = inp.readlines()

howManyCols = len(lines[0])
howManyLines = len(lines)

def check_horizontal(lis, lin, col):
    try:
        word=""
        for i in range(0,4):
            word += lis[lin][col+i]
        return word in ("XMAS","SAMX")
    except:
        return False
    
def check_vertical(lis, lin, col):
    try:
        word=""
        for i in range(0,4):
            word += lis[lin+i][col]
        return word in ("XMAS","SAMX")
    except:
        return False
    
def check_diagonal(lis, lin, col):
    try:
        word = ""
        for i in range(0,4):
            word += lis[lin+i][col+i]
        return word in ("XMAS","SAMX")
    except:
        return False

def check_antidiagonal(lis, lin, col):
    try:
        word =""
        for i in range(0,4):
            a = col-i
            if a<0:
                raise ValueError
            word += lis[lin+i][a]
        return word in ("XMAS","SAMX")
    except:
        return False

def x_mas(lis,lin,col):
    # 1 . .    . . 3
    # . 2 .    . 2 .
    # . . 3    1 . .
    word1=""
    word2=""
    for i in range(-1, 2):
        word1+=lis[lin+i][col+i]
        word2+=lis[lin-i][col+i]
    return (word1, word2)